src/entities/mellowLpVault.ts

```diff
@@ -0,0 +1,437 @@
+import timestampWadToDateTime from '../utils/timestampWadToDateTime';
+import { getGasBuffer, MaxUint256Bn, TresholdApprovalBn } from '../constants';
+
+import { abi as VoltzVaultABI } from '../ABIs/VoltzVault.json';
+import { abi as Erc20RootVaultABI } from '../ABIs/Erc20RootVault.json';
+import { abi as Erc20RootVaultGovernanceABI } from '../ABIs/Erc20RootVaultGovernance.json';
+import { abi as MarginEngineABI } from '../ABIs/MarginEngine.json';
+import { abi as BaseRateOracleABI } from '../ABIs/BaseRateOracle.json';
+import { abi as IERC20MinimalABI } from '../ABIs/IERC20Minimal.json';
+
+export type MellowLpVaultArgs = {
+  voltzVaultAddress: string;
+  erc20RootVaultAddress: string;
+  erc20RootVaultGovernanceAddress: string;
+  provider?: providers.Provider;
+};
+
+class MellowLpVault {
+  public readonly voltzVaultAddress: string;
+  public readonly erc20RootVaultAddress: string;
+  public readonly erc20RootVaultGovernanceAddress: string;
+  public readonly provider?: providers.Provider;
+
+  public readOnlyContracts?: {
+    marginEngine: Contract;
+    token: Contract;
+    rateOracle: Contract;
+    voltzVault: Contract;
+    erc20RootVault: Contract;
+    erc20RootVaultGovernance: Contract;
+  };
+
+  public writeContracts?: {
+    token: Contract;
+    voltzVault: Contract;
+    erc20RootVault: Contract;
+  };
+
+  public signer?: Signer;
+  public maturity?: string;
+  public protocolId?: number;
+
+  public vaultAccumulative?: number;
+  public vaultCap?: number;
+  public vaultExpectedApy?: number;
+
+  public userDeposit?: number;
+  public userExpectedCashflow?: number;
+  public userWalletBalance?: number;
+
+  public userAddress?: string;
+
+  public vaultInitialized = false;
+  public userInitialized = false;
+
+  public constructor({
+    erc20RootVaultAddress,
+    erc20RootVaultGovernanceAddress,
+    voltzVaultAddress,
+    provider,
+  }: MellowLpVaultArgs) {
+    this.erc20RootVaultAddress = erc20RootVaultAddress;
+    this.erc20RootVaultGovernanceAddress = erc20RootVaultGovernanceAddress;
+    this.voltzVaultAddress = voltzVaultAddress;
+    this.provider = provider;
+  }
+
+  descale = (amount: BigNumberish, decimals: number): number => {
+    return Number(ethers.utils.formatUnits(amount, decimals));
+  };
+
+  scale = (amount: number): BigNumber => {
+    return ethers.utils.parseUnits(amount.toString(), this.tokenDecimals);
+  };
+
+  // NEXT: to offload this to subgraph
+  vaultInit = async (): Promise<void> => {
+    if (this.vaultInitialized) {
+      console.log('The vault is already initialized');
+      return;
+    }
+
+    if (isUndefined(this.provider)) {
+      console.log('Stop here... No provider provided');
+      return;
+    }
+
+    const voltzVaultContract = new ethers.Contract(
+      this.voltzVaultAddress,
+      VoltzVaultABI,
+      this.provider,
+    );
+    console.log('voltz vault address:', this.voltzVaultAddress);
```

Comment:
do we want to keep console logs?

Reply:
I think I'm going to keep them during the QA phase.